File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	
	static int N, sol;
	static int[][] map;
	static boolean[][] visited;
	static int maxScore;
	
	static int[] dr = {0,0,1,-1};
	static int[] dc = {1,-1,0,0};

	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		N = Integer.parseInt(br.readLine());
		
		map = new int[N][N];
		for (int r=0; r<N ; r++) {
			st = new StringTokenizer(br.readLine());
			for (int c=0; c<N; c++) {
				map[r][c] = Integer.parseInt(st.nextToken());
				maxScore = Math.max(maxScore, map[r][c]);
			}
		}
		
		for (int s=0; s<=maxScore; s++) {
			search(s);
		}
		
		System.out.println(sol);
		
		
	}
	
	private static void search(int limit) {
		
		int cnt = 0;
		
		visited = new boolean[N][N];
		
		for (int r=0; r<N; r++) {
			for (int c=0; c<N; c++) {
				if (map[r][c] > limit) continue;
				visited[r][c] = true;
			}
		}
		
		for (int r=0; r<N; r++) {
			for (int c=0; c<N; c++) {
				if (visited[r][c]) continue;
				
//				for (int i=0; i<N; i++) {
//					System.out.println(Arrays.toString(visited[i]));
//				}
//				System.out.println();
				
				BFS(r,c,limit);
				cnt++;
			}
		}
		
		sol = Math.max(sol, cnt);
		
	}
	
	private static void BFS(int startR, int startC, int limit) {
		
		Queue<int[]> q = new ArrayDeque<>();
		q.add(new int[] {startR, startC});
		visited[startR][startC] = true;
		
		while (!q.isEmpty()) {
			int[] coord = q.poll();
			int r = coord[0];
			int c = coord[1];
			for (int d=0; d<4; d++) {
				int nr = r+dr[d];
				int nc = c+dc[d];
				if (nr >= N || nc >= N || nr < 0 || nc < 0 || visited[nr][nc]) continue;
				q.add(new int[] {nr, nc});
				visited[nr][nc] = true;
			}
			
		}
		
	}

}

File: 백준/Silver/2468. 안전 영역/안전 영역.py
from collections import deque

ds = [[1, 0], [-1, 0], [0, 1], [0, -1]]

def bfs():
    visited = [[0] * N for _ in range(N)]

    cnt = 0         # 영역의 개수
    for sr in range(N):     # 시작지점 설정
        for sc in range(N):
            if arr[sr][sc] > h and visited[sr][sc] == 0:   # h보다 크고(즉, h일때 안잠기고) 방문하지 않았다면
                q = deque()     # 각 시작지점마다 q를 초기화
                
                visited[sr][sc] = 1     # 시작지점 방문처리, 큐 삽입
                q.append([sr, sc])
                cnt += 1                # 영역 개수 추가
                while q:
                    cr, cc = q.popleft()
                    for dr, dc in ds:
                        nr, nc = cr + dr, cc + dc
                        if 0 <= nr < N and 0 <= nc < N:     # 범위 내에 있고
                            if arr[nr][nc] > h and visited[nr][nc] == 0:
                                visited[nr][nc] = 1
                                q.append([nr, nc])

    return cnt


N = int(input())
arr = [list(map(int, input().split())) for _ in range(N)]

ans = 0
for h in range(100):  # 높이가 100인 좌표가 있다면, h:100까지도 고려해야함
    ans = max(ans, bfs())    # 높이가 h일때의 최대영역개수
print(ans)
